### DISCUSSION TITLE
python-oracledb: cannot mTLS connect to Oracle DB on Amazon RDS service in Thin mode

### DISCUSSION THREAD

**Discussion by FrancoisNoyez** [oracle/python-oracledb#62] [Q&A] [answered]
**Context**
I want to implement python-based Cloud Functions on Google Cloud Plateform. Those Cloud Functions will need to connect to our Amazon RDS Oracle DB.  
Until recently this was mostly very difficult / impossible, since using cx_Oracle would require the client software to be installed in the environnement set up for the Cloud Function, and GCP would not let us easily do that (I did try at length...).
But oracledb allows to implement that, with it's Thin mode.

I can't develop and test directly on my local computer, since the way I access the DB is different there:
- when I use my computer, I need to connect to the DB through a ssh tunnel opened through a server whose IP is whitelisted for the RDS OracleDB
- for the Cloud Functions it will be different: they will be attached to a VPN such that the network's elements in it are whitelisted for the RDS OracleDB

I'm already responsible for the maintenance of a project which uses jobs done through pods created in that VPN, so what I do to dev is that I manually create such a pod, and tests things whitin it. The pods for this projet are created through the use of a Docker image which has been prepared to contain the Oracle instantclient software, and the program uses cx_Oracle to connect. To achieve the connection we need to use mTLS, through the use of a .sso wallet file.

I read the [doc](https://python-oracledb.readthedocs.io/en/latest/user_guide/connection_handling.html#mutual-tls-mtls-connection-to-oracle-autonomous-database) in order to know how to replicate that when using Thin mode of python-oracledb, but I was unable to do so.
The initial difficulty was about getting the .pem file (I don't have knowledge / experience regarding encryption, I just try to follow what the docs tell me). I finally got one from [here](https://lightsail.aws.amazon.com/ls/docs/en_us/articles/amazon-lightsail-download-ssl-certificate-for-managed-database), I took the rds-ca-2019-root.pem file.
I renamed it to ewallet.pem, and put it in a directory whose path I give when I try to connect.
But when I test this, the connection fails, with an error message which does not say anything useful regarding what went wrong.
  
  
**Code**
```python
# encoding: utf-8

import oracledb
import os

username = os.environ["ORACLE_USER"]
password = os.environ["ORACLE_PASSWORD"]
dsn = """(DESCRIPTION=(ADDRESS=(PROTOCOL=TCPS)(HOST=DB_NAME.ap-southeast-1.rds.amazonaws.com)(PORT=2484))(CONNECT_DATA=(SID=XXXXX)(SERVICE_NAME=XXXXX)))"""
lib_dir = "/opt/instantclient_19_3"
config_dir = "/opt/oracle/network/admin"
wall_loc = "/code/wallet2"
wall_pw = None
# wall_pw = ""

use_thick_mode = False

kwargs = {
    "config_dir": config_dir,
    "wallet_location": wall_loc,
    "wallet_password": wall_pw,
}
if use_thick_mode:
    oracledb.init_oracle_client(lib_dir=lib_dir)
    kwargs = {}

with oracledb.connect(
    user=username,
    password=password,
    dsn=dsn,
    **kwargs,
) as ocon:
    with ocon.cursor() as cursor:
        sql = """select '['||systimestamp||'] Connectivity OK' from dual"""
        for r in cursor.execute(sql):
            print(r[0])
```
  
  
**Error**
```sh
root@POD_NAME:/code# python test_oracledb_connect2.py 
Traceback (most recent call last):
  File "src/oracledb/impl/thin/connection.pyx", line 97, in oracledb.thin_impl.ThinConnImpl._connect_with_address
  File "src/oracledb/impl/thin/protocol.pyx", line 145, in oracledb.thin_impl.Protocol._connect_phase_one
  File "src/oracledb/impl/thin/protocol.pyx", line 279, in oracledb.thin_impl.Protocol._connect_tcp
  File "src/oracledb/impl/thin/crypto.pyx", line 118, in oracledb.thin_impl.get_ssl_socket
ssl.SSLError: [SSL] PEM lib (_ssl.c:3932)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "test_oracledb_connect2.py", line 31, in <module>
    **kwargs,
  File "/usr/local/lib/python3.7/site-packages/oracledb/connection.py", line 1000, in connect
    return conn_class(dsn=dsn, pool=pool, params=params, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/oracledb/connection.py", line 128, in __init__
    impl.connect(params_impl)
  File "src/oracledb/impl/thin/connection.pyx", line 294, in oracledb.thin_impl.ThinConnImpl.connect
  File "src/oracledb/impl/thin/connection.pyx", line 184, in oracledb.thin_impl.ThinConnImpl._connect_with_params
  File "src/oracledb/impl/thin/connection.pyx", line 157, in oracledb.thin_impl.ThinConnImpl._connect_with_description
  File "src/oracledb/impl/thin/connection.pyx", line 109, in oracledb.thin_impl.ThinConnImpl._connect_with_address
  File "/usr/local/lib/python3.7/site-packages/oracledb/errors.py", line 103, in _raise_err
    raise exc_type(_Error(message)) from cause
oracledb.exceptions.OperationalError: DPY-6005: cannot connect to database. Connection failed with "[SSL] PEM lib (_ssl.c:3932)"
```

  

**What I've tried**
I've used the .pem file that I downloaded in order to create a fill a new .sso wallet file, and I used it either with cx_Oracle or with the Thick mode of oracledb (cf the code snippet above), and I was able to connect to the Oracle DB without issue.

When I remove the ewallet file from the expected location, there is an error message that says that the file is missing:
```sh
oracledb.exceptions.ProgrammingError: DPY-2018: cannot connect to database. Wallet file /code/wallet2/ewallet.pem was not found
```
When I create an empty cwallet.pem file, and put it where oracledb is expecting it when using Thin mode, there is an error message that says that the file does not contain a certificate:
```sh
oracledb.exceptions.OperationalError: DPY-6005: cannot connect to database. Connection failed with "[X509] no certificate or crl found (_ssl.c:4140)"
```
So I'm confident that oracledb is able to find the file, load its content, and search for a certificate.

I'm not too sure what to do regarding the wallet_password value, since I have no info about that from the source of the .pem file: the Amazon site does not say anything about that, so I'm assuming that I don't need to specify anything?
Just to be sure, I've browsed the web, and saw tutos which explain how to convert from the .sso wallet file back to the .pem file, and at some point we can specify passwords. I've done that, but even with rebuilt .pem file, and me specifying the correct password, the error stays the same.
For information, the script described in the [doc](https://python-oracledb.readthedocs.io/en/latest/user_guide/connection_handling.html#createpem) to convert from .p12 to .pem did not work for me, so there is probably something to fix there:
```sh
francois@francois-laptop#:~/ python create_pem.py --wallet-password "XXXXX" ./
Traceback (most recent call last):
  File "create_pem.py", line 45, in <module>
    f.write(private_key.private_bytes(Encoding.PEM, PrivateFormat.PKCS8,
AttributeError: 'NoneType' object has no attribute 'private_bytes'
```

Finally, I've also tried to test this using the 1.0.0, 1.0.1, and 1.0.2 version of oracledb (I use 1.0.3 when I ask to install the library without specifying the version), but I always got the same underlying error: "[SSL] PEM lib (_ssl.c:3932)".

**Question**
- What does this error mean? Would it be possible to update the library so that the error can be explicit, like it is when the file is missing or when it's empty?
- What is the root cause of the error? Is there something specific to apply to the .pem file beforehand? Seems like the use of the certificate of this file is enough to make things work when using Thick mode, though... In that case, is there an issue with the current version of the library for some cases of .pem file? Why would it work for other people but not me, though : / ?

**Comment by anthony-tuininga** [comment]
Hi, @FrancoisNoyez, thanks for the detailed descriptions of what you did. If you are using mTLS the `ewallet.pem` file is needed, and, unless you have stored it unecrypted, you also need to supply the `wallet_password` value. That will be the value that is required to decrypt the encrypted private key and certificates that are stored in the `ewallet.pem` file. If you supply the wrong value (or no value) then you will get the `"[SSL] PEM lib"` error. Yes, the error is terrible, but its not one that we control, unfortunately!

Regarding the conversion from .p12 to .pem -- it looks like you don't have a private key in the .p12 file. Where did you get that file from? For Oracle Cloud databases, the wallet is downloaded from console and is a zip file containing these files:

```
cwallet.sso  ewallet.pem   ojdbc.properties  sqlnet.ora    truststore.jks
ewallet.p12  keystore.jks  README            tnsnames.ora
```

Thick mode doesn't look at `ewallet.pem` but looks at `cwallet.sso` instead. Where did you get that file from? Does it have an accompanying `ewallet.p12` or `ewallet.pem`?

**Reply by FrancoisNoyez** [reply]
Hi Anthony,

Thank you for your answer.
Ok, so you are saying that this error means that there is an issue with the `wallet_password` value passed to the library, compared to the one which is expected to make things work.
Hm, that's a shame, regarding the explainability of the error. Isn't there a way to capture it, and to wrap it with something better, within oracledb?

Yes, I think the higher level issue is the way for me to get the .pem file. The one I used, I got it from [here](https://lightsail.aws.amazon.com/ls/docs/en_us/articles/amazon-lightsail-download-ssl-certificate-for-managed-database). I'm under the impression that there is no password (even no private key?) associated to them / that those are destined to be used in a general way?
What is strange is that, when I embedd this .pem certificate file in a cwallet.sso file, using the following commands
```sh
orapki wallet create -wallet . -auto_login_only
orapki wallet add -wallet . -trusted_cert -cert rds-ca-2019-root.pem -auto_login_only
```
, and then I use this `cwallet.sso` file in Thick mode, things work properly, as I've described in my post. Is it because, when the certificate is embedded in the wallet, when the time comes to verify things using the certificate, there is no need for such a password, somehow?

For reference, in order to have access to the `orapki` command, I used a Docker image containing the Oracle software (cf [here](https://github.com/fuzziebrain/docker-oracledb21c-xe)).

As for how the original `cwallet.sso` file which is used in the current jobs has been created in the first place, I don't know, as the person who developped that projet which I'm maintaining has left our organization since then. But as I've shown, I know how to recreate one which works, so for now I'm assuming that this person did something similar.

I don't have access to the Oracle Cloud console for this database that I want to connect to. I'll have to ask the person in my organization who has such an access. Assuming I get all those files through a .zip archive from using the Oracle Cloud console, how would I get the password? Would it be in the README fiel in archive, for instance?

**Reply by cjbj** [reply]
As part of the 'connection' download process from the Oracle Cloud console of the wallet.zip file there is a prompt for the wallet password.  Whoever gives you this wallet would need to separately give you that password.  An ewallet.pem file was included in the wallet.zip downloads about the time that python-oracledb 1.0 was released.  Earlier zip files won't have a PEM file.  But if your database is on Amazon, I don't know what is done there.

**Reply by FrancoisNoyez** [reply]
Ok, I see, thanks. Then I guess the only way for me to make this thing work is to get the .pem file through this process of downloading the archive from the Oracle Cloud console.
I'll see to contact the appropriate person within my organization.

Still odd that using the .pem file that I used here work when embedded inside a .sso wallet, though...

I'll come back here once I've got the correct .pem file, and confirmed that it makes things work.

**Reply by anthony-tuininga** [reply]
Have you tried without the `wallet_location` and `wallet_password` parameters? Relatively recently support for one-way TLS was added to the database -- not sure if that made its way to the Amazon hosted databases? Note that cwallet.sso can contain the wallet password and the thick driver knows how to extract that -- but the thin driver doesn't understand that. Without an understanding of what Amazon is doing I can't really answer directly what you need to do. :-(

**Reply by FrancoisNoyez** [reply]
Hi Anthony,

Yes, I have.
When I set both variables to None, I get the following error:
```sh
oracledb.exceptions.OperationalError: DPY-6005: cannot connect to database. Connection failed with "[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate in certificate chain (_ssl.c:1091)"
```
When I use `PROTOCOL=TCP` instead of `PROTOCOL=TCPS` inside the dsn value, and still setting both variables to None, I get the following error:
```sh
oracledb.exceptions.DatabaseError: DPY-4011: the database or network closed the connection
[Errno 104] Connection reset by peer
```
which means the DB I'm trying to access really requires people to authenticate when trying to do so.

Regarding the fact that the .sso wallet works when using Thick mode, the thing is, when I create one from embedding the .pem file, I never set any kind of password up. So I'm really puzzled about why it works in that case.

No problem, I understand. Seems like oracledb really expects to use a .pem file provided through the use of the Oracle Cloud console, like what is explained in the doc, so I just have to get it like that. Since I could not get it right away, and because the person within my organization who I first contacted redirected me to the page of the Amazon website where those "global" config .pem file are available, I thought that I could get away with using them. But alas, no : /

I'll warn you once I've gotten hold of those files downloaded through the Oracle Cloud console.

**Reply by anthony-tuininga** [reply]
Your database is actually using a self-signed certificate? In production? Interesting! That suggests that one-way TLS is actually being used -- which makes life possibly less complicated. :-) That might also explain why just the certificate is needed for the thick driver and not a private key as well.

You need to get the OS to trust the certificate (in the same way that you would get the OS to trust a self-signed certificate from a web server). This differs from platform to platform so I can't give you exact instructions -- but hopefully Google can help there! If you can complete this step the connection may succeed (if one-way TLS is indeed in place).

**Reply by FrancoisNoyez** [reply]
I see.
Now I make more sense of [this blog article](https://www.dbaglobe.com/2022/07/using-python-oracledb-to-connect-rds.html?sc=1662139105024#c7682417952820709591) that I had found during earlier investigations, and that I had tried to use. Seems like getting the OS to trust the certificate is the point of this article. I tried to replicate it within the Docker instance of my current jobs, but with little success. That said, even if I had succeeded, since the end goal is to make things work within a CloudFunction, and since I probably can't apply the needed commands in that context (I did try to use subprocesses, when I tried to install the Oracle software, but it did not work), it's probably moot.

**Reply by anthony-tuininga** [reply]
Yes, that article looks exactly like what you need to do -- so the fact that it doesn't work for you makes life complicated! I'm not sure what else to tell you at this point. If you are able to build python-oracledb yourself, you can modify the following lines:

https://github.com/oracle/python-oracledb/blob/main/src/oracledb/impl/thin/crypto.pyx#L127-L136

Specifically, you can pass a value to `wallet_location` but *comment out* the line that calls `load_cert_chain()` since that expects a private key. If that works for you, great! Then we can discuss how to make this possible without hacking the code. :-)

**Reply by FrancoisNoyez** [reply]
Hi Anthony,
  
Great idea, it worked! I commented this line out, rebuilt the package, and used it, and the code was able to connect to the DB!  
  
Seems like indeed the issue is that, the way the library is implemented, if a non-null 'wallet_location' value is provided, then the library expects that it will contain a password / a private key to be compared to the value provided by 'wallet_password'.  Here things break down because there is no such password / private key to be extracted from the .pem file.  
  
One way to improve the library that I would suggest would be to test for the existence of this in the .pem file, and compare that to the value provided by 'wallet_password' (with null value being interpreted as wanting to use the certificate for one-way TLS), and then proceed accordingly with the rest of the code, or at least raising explicit exceptions if there is a mismatch.  
Or to have an additional boolean parameter such as 'use_one_way_tls', and decide whether or not to "carry out the search for password / private key and compare it to the 'wallet_password' value" accordingly, and still with the raising of explicit exception if there is an issue.  
  
Anyway, thank you for your time, with that I have a way to move forward with my project! I'll still want to fetch the .pem file generated through the Oracle Cloud console, for reference, and just in case. For now I plan to not resolve the question / mark this post as answer until I have gotten it, and verified that all things work properly, if that's ok with you. Don't worry, I won't forget.

**Reply by FrancoisNoyez** [reply]
Ok, so apparently the people in my organisation in charge of managing our Oracle DB do not have access to the Oracle console, as apparently this requires to be the admin of the server where the DB is hosted, and we are not.  
And I can't install the custom version of the library inside the CloudFunction.  
So the easier solution for us would be that "python-oracledb" be updated to allow to use .pem file which do not contain a private key / which are unencrypted.  
Since I have made it work with my custom modification, I will want to share it here, by creating a PR. According to the contribution guidelines, I first need to create a github issue, so that's what i'll do.

**Comment by cjbj** [comment, marked as answer]
For the record, this was enhanced in python-oracledb 1.2.  See https://github.com/oracle/python-oracledb/issues/65

**Reply by FrancoisNoyez** [reply]
Got it, indeed things work now; thank you : ) !
